pkg/resource/claim_scheduling_reconciler.go

```diff
@@ -0,0 +1,157 @@
+func WithJitterer(j Jitterer) ClaimSchedulingReconcilerOption {
+	return func(r *ClaimSchedulingReconciler) {
+		r.jitter = j
+	}
+}
+
+// NewClaimSchedulingReconciler returns a ClaimSchedulingReconciler that
+// schedules resource claims to a resource class that matches their class
+// selector.
+func NewClaimSchedulingReconciler(m manager.Manager, of ClaimKind, to ClassKind, o ...ClaimSchedulingReconcilerOption) *ClaimSchedulingReconciler {
+	nc := func() Claim { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Claim) }
+
+	// Panic early if we've been asked to reconcile a claim or resource kind
+	// that has not been registered with our controller manager's scheme.
+	_ = nc()
+
+	r := &ClaimSchedulingReconciler{
+		client:    m.GetClient(),
+		newClaim:  nc,
+		classKind: to,
+		jitter: func() {
+			random := rand.New(rand.NewSource(time.Now().UnixNano()))
+			time.Sleep(time.Duration(random.Intn(200)) * time.Millisecond)
+		},
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a resource claim by using its class selector to select and allocate
+// it a resource class.
+func (r *ClaimSchedulingReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("Reconciling", "controller", claimSchedulingControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), claimSchedulingReconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// There could be several controllers racing to schedule this claim. If it
+	// was scheduled since we were queued then another controller won and we
+	// should abort.
+	if claim.GetClassReference() != nil {
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	classes := &unstructured.UnstructuredList{}
+	classes.SetGroupVersionKind(r.classKind.List())
+
+	if err := r.client.List(ctx, classes, client.MatchingLabels(claim.GetClassSelector().MatchLabels)); err != nil {
+		// Claim scheduler controllers don't update the synced status because
+		// no one scheduler has the full view of whether the process failed or
+		// succeeded. It's possible another controller can successfully set a
+		// class even though we can't, so it would be confusing to mark this
+		// claim as failing to be reconciled. Instead we return an error - we'll
+		// be requeued but abort immediately if the claim was scheduled.
+		return reconcile.Result{}, errors.Wrap(err, errListClasses)
+	}
+
+	if len(classes.Items) == 0 {
+		// None of our classes matched the class selector.
+		// There's nothing for us to do.
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	random := rand.New(rand.NewSource(time.Now().UnixNano()))
```

Comment:
I like that we don't just use the first one we find, reminds me of [this property](https://nathanleclaire.com/blog/2014/04/27/a-surprising-feature-of-golang-that-colored-me-impressed/) of go.

Reply:
Do you think there would be much value in doing this in the defaulting controller too?

I originally had that randomly picking one default when multiple classes were annotated as default (pretty much the same logic as the scheduling controller), but then I figured it was acceptable (and simpler) to just pick the first class given that the existence of multiple default classes is a misconfiguration.

That said, it would not be that hard to pick one randomly (just copy the logic from the scheduling controller), and doing so could conceivably be useful to some users (i.e. if you wanted to randomly spread unopinionated claims across N different "default" resource classes)/

Reply:
I like the idea of picking a random one because it removes the ability for a user to either:

1. Have multiple defaults configured and not fix it because they know the one they want is going to get picked (not sure if the `List()` returns predictable order?).
2. Have multiple defaults configured and not be aware of it because the consistently get the same one.

TLDR: I would support doing this in the defaulting controller :)

Reply:
I think I agree, but I might just add a TODO for now and address this in a future PR.